File: Data Structures & Algorithms/minimum-window-with-characters/submission-1.java
class Solution {
    public String minWindow(String s, String t) {
        String res = "";

        HashMap<Character, Integer> smap = new HashMap<>();
        HashMap<Character, Integer> tmap = new HashMap<>();

        int matches = 0;
        int neededMatches = 0;

        for (int i = 0; i < t.length(); i++) {
            tmap.put(t.charAt(i), tmap.getOrDefault(t.charAt(i), 0) + 1);
            if (tmap.get(t.charAt(i)) == 1) {
                neededMatches++;
            }
        }

        int left = 0;
        int right = 0;

        while (right < s.length()) {
            char cur = s.charAt(right);
            smap.put(cur, smap.getOrDefault(cur, 0) + 1);

            if (smap.get(cur) == tmap.get(cur)) {
                matches++;
            }

            while (matches == neededMatches) {
                // Update res
                String sub = s.substring(left, right + 1);
                if (sub.length() < res.length() || res.isEmpty()) {
                    res = sub;
                }

                char leftChar = s.charAt(left);

                // Contract Window
                smap.put(leftChar, smap.get(leftChar) - 1);
                if (tmap.containsKey(leftChar) && smap.get(leftChar) < tmap.get(leftChar)) {
                    matches--;
                }
                left++;
            }

            right++;
        }

        return res;
    }
}

File: Data Structures & Algorithms/decode-ways/submission-1.py
class Solution:
    def numDecodings(self, s: str) -> int:
        dp = [-1] * len(s)

        def valid_decoding(i, j):
            if i < 0 or j > len(s) or i >= j:
                return False
            sub = s[i:j]
            if len(sub) == 1:
                return sub in ["1", "2", "3", "4", "5", "6", "7", "8", "9"]
            return sub[0] in ["1", "2"] and int(sub) <= 26
        
        dp[0] = 1 if valid_decoding(0, 1) else 0
        if len(s) > 1:
            total = 0
            if dp[0] != 0:
                if valid_decoding(1, 2):
                    total += 1
                if valid_decoding(0, 2):
                    total += 1
            dp[1] = total

        i = 2
        while i < len(s):
            total = 0
            if valid_decoding(i, i + 1):
                total += dp[i - 1]
            if valid_decoding(i - 1, i + 1):
                total += dp[i - 2]
            dp[i] = total
            i += 1
        print(dp)
        return dp[len(s) - 1]
            